docs/architecture/adr-038-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,86 @@
+
+Draft
+
+## Abstract
+
+Currently in the SDK, light clients are part of Cosmos SDK codebase. What this means that, anytime a blockchain built 
+upon Cosmos SDK need to add support for new light client, Cosmos SDK need to be modified and all validator nodes of that 
+blockchain need to be updated with custom version of Cosmos SDK. 
+
+To remedy these shortcomings, we are proposing a WASM VM to host light client bytecode, which allows easier upgrading of 
+existing light clients as well as adding support for new light clients without requiring a code release and corresponding 
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under 
+`/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the event of security 
+issue or consensus update, we need to modify the codebase and integrate it in *numerous* places. In addition, individual 
+blockchains built upon Cosmos SDK need to update all their validator nodes to latest version to add support for this 
+light client. This entire process is tedious and time consuming. In the event a blockchain wants to add support for niche 
+light client, it need to fork Cosmos SDK and modify codebase to add support for it. This creates overhead of maintaining 
+that fork against mainstream Cosmos SDK release.
+
+We are proposing simplifying this workflow by integrating a WASM light client module which makes adding support for 
+new light client a simple transaction. The light client bytecode, written in Wasm-compilable Rust runs inside a Wasmer 
+VM. The Wasm light client submodule exposes a proxy light client interface that routes incoming messages to the 
+appropriate handler function, inside the Wasm VM for execution.
+
+## Decision
+
+We decided to use WASM light client module as a generic light client which will interface with the actual light client 
+uploaded as WASM bytecode. This will require changing client selection method to allow any client if the client type 
+has prefix of `wasm/`.
+
+```go
+// IsAllowedClient checks if the given client type is registered on the allowlist.
+func (p Params) IsAllowedClient(clientType string) bool {
+	for _, allowedClient := range p.AllowedClients {
+		if allowedClient == clientType {
+			return true
+		}
+	}
+	return false || isWASMClient(clientType)
+}
+```
+
+Inside Wasm light client `ClientState`, appropriate Wasm bytecode will be executed depending upon `ClientType`.
+
+```go
+func (cs ClientState) Validate() error {
+    wasmRegistry = getWASMRegistry()
+	clientType := cs.ClientType()
+    codeHandle := wasmRegistry.getCodeHandle(clientType)
+    return codeHandle.validate(cs)
+}
+```
+
+To upload new light client, user need to create a transaction with Wasm byte code and unique client id, which will be 
+processed by IBC Wasm module.
+
+```go
+func (k Keeper) UploadLightClient (WasmCode: []byte, id: String, description: String) {
```

Comment:
I don't think we can safely allow users to upload light client byte code? What stops a user from uploading malicious byte code which just verifies every packet as valid?

Reply:
Is `validateAndStoreCode` implying verification of who is updating the client? This would be good information to add to the ADR

Reply:
This is fine. It would basically be allowing users to submit new client implementations, but it is still incumbent on relayers to create the right connections with the right clients.

The client state struct stored in IBC store should include the wasm byte code, or at least a hash of wasm byte code.

This will then allow a chain to check that the counterparty is executing the appropriate client state code.

**NOTE:** This will enforce either one or a specific subset of implementations for each client types.

Reply:
Original idea was to allow anyone to upload new light client on chains with non-zero gas fees provided He/she is comfortable with paying significant gas costs. But, unless a relayer/client creator does not use that new version, it will be just extra storage in blockchain state.

In future, we can also involve a governance process where light client is disabled initially and is only enabled when majority votes yes during governance process.